### ISSUE TITLE
[ICE] crash on function initializers

### ISSUE THREAD

**Issue by jyn514** [jyn514/rcc#259]
**Expected behavior**
rcc gives a semantic error

**Actual Behavior**
rcc crashes

**Code**

```c
int i, f() = {1};
The application panicked (crashed).
Message:  not yet implemented: type checking for aggregate initializers of type int()
Location: src/parse/decl.rs:1752
```

The following code also crashes: `int i, a()[] = {1};`

<details><summary>Backtrace</summary>

```
 5: rcc::parse::decl::<impl rcc::data::types::Type>::type_at::hfade243071ad41ae
    at src/parse/decl.rs:1752
 6: rcc::parse::decl::<impl rcc::parse::Parser<I>>::aggregate_initializer::hf7414657fac40f63
    at src/parse/decl.rs:1317
 7: rcc::parse::decl::<impl rcc::parse::Parser<I>>::initializer::hfb3a30bf6fc27967
    at src/parse/decl.rs:1272
 8: rcc::parse::decl::<impl rcc::parse::Parser<I>>::init_declarator::h881d74f468159736
    at src/parse/decl.rs:308
 9: rcc::parse::decl::<impl rcc::parse::Parser<I>>::declaration::hf346af91832a5bf5
    at src/parse/decl.rs:159
10: <rcc::parse::Parser<I> as core::iter::traits::iterator::Iterator>::next::{{closure}}::h3e16e32b7ef142c5
    at src/parse/mod.rs:147
11: core::option::Option<T>::or_else::h13abe77b2f91e609
    at /rustc/73528e339aae0f17a15ffa49a8ac608f50c6cf14/src/libcore/option.rs:751
12: <rcc::parse::Parser<I> as core::iter::traits::iterator::Iterator>::next::hf86c57aa21e3f2c9
    at src/parse/mod.rs:125
13: <&mut I as core::iter::traits::iterator::Iterator>::next::hbad70cf06cf79949
    at /rustc/73528e339aae0f17a15ffa49a8ac608f50c6cf14/src/libcore/iter/traits/iterator.rs:2990
14: rcc::compile::h75cd05a7ff149c39
    at src/lib.rs:172
15: rcc::real_main::h6732a6d54c22fad0
    at src/main.rs:105
16: rcc::main::h285e4f67e76a9791
    at src/main.rs:171
```

</details>

**Comment by jyn514** [comment]
Mentoring instructions:

- In [`src/parse/decl.rs:init_declarator`](https://github.com/jyn514/rcc/blob/master/src/parse/decl.rs#L290), after parsing a declarator, check for whether the declarator is a function. If so, do not parse an initializer (just leave the tokens unconsumed, the next call to declarator will give a syntax error).

**Comment by jyn514** [comment]
Alternatively, you could still call `initializer` but set the type to Type::Error. I don't think this is the appropriate solution though because it turns a syntax error into a semantic one.

**Comment by jyn514** [comment]
Actually I take it back, I think it should be a semantic error since it meets the `declarator "=" initializer` syntax. So then the appropriate thing to do is give an error (adding an new variant to [`SemanticError`](https://github.com/jyn514/rcc/blob/master/src/data/error.rs#L91)
 if necessary) and then set the type of the declarator to Type::Error.

**Comment by jyn514** [comment]
Fixed at some point (possibly in #370?)
